### PR TITLE
Create initial_state tensor filled with zeros without use of K.zeros

### PR DESCRIPTION
**Pull request by yosinski**
`Recurrent.get_initial_states` contains some awkward machinations to get an `initial_state` vector of the appropriate size filled with zeros, which can only be done at graph execution time because the batch size is needed.

The current implementation uses a call to `K.zeros`, which when using the TensorFlow backend creates a Variable.This is not desirable, because
 * By default all Variable objects will be collected and trained by TF optimizers, wasting time computing the (zero) gradients, and
 * By default all Variable objects will be serialzed and saved when using the `tf.train.Saver`, wasting space.

The fix is simple: use only `K.zeros_like`, which creates only operations in the graph, not extra Variables.

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -190,9 +190,9 @@ def get_constants(self, x):
     def get_initial_states(self, x):
         # build an all-zero tensor of shape (samples, output_dim)
         initial_state = K.zeros_like(x)  # (samples, timesteps, input_dim)
-        initial_state = K.sum(initial_state, axis=1)  # (samples, input_dim)
-        reducer = K.zeros((self.input_dim, self.output_dim))
-        initial_state = K.dot(initial_state, reducer)  # (samples, output_dim)
+        initial_state = K.sum(initial_state, axis=(1,2))  # (samples,)
```
Comment (fchollet) [inline keras/layers/recurrent.py]: PEP8: space after the comma.

**Comment by fchollet** [comment]
I think it should be possible to instead build an integer tensor `shape` using `K.shape(x)` (symbolic) then create a new variable using the symbolic shape tensor.

File: keras/layers/recurrent.py (hunk 1 above)
Comment (yosinski) [inline keras/layers/recurrent.py]: Fixed.

**Comment by yosinski** [comment]
Creating an integer `shape` tensor containing the target shape of `initial_states` using `K.shape(x)` should be possible. However, using it to create `initial_states` would then require use of `K.zeros`, which creates a `tf.Variable` when using the TF backend. Creating extraneous `Variable` objects is probably Not Desired (per above).

The issue arises because the TF and Keras `zeros` functions behave slightly differently. If we use TF to produce tensors filled with zeros via either `zeros_like` or `zeros`, both produce ops that output tensors without creating any variables:

```
In [1]: import tensorflow as tf
In [2]: import keras.backend as K
In [3]: reference = tf.placeholder('float32', shape=(2,3,4))
```

Use TF to make zeros tensors via both methods:

```
In [4]: zer_like = tf.zeros_like(reference)
In [5]: zer = tf.zeros((2,3,4))
```

And note that both methods result in a Tensor, not a Variable:

```
In [6]: zer_like
Out[6]: <tf.Tensor 'zeros_like:0' shape=(2, 3, 4) dtype=float32>
In [7]: zer
Out[7]: <tf.Tensor 'zeros:0' shape=(2, 3, 4) dtype=float32>
In [8]: print tf.get_default_graph().get_collection('variables')
[]
```

However, if we call the Keras wrappers, the behavior is different:

```
In [9]: k_zer_like = K.zeros_like(reference)
In [10]: k_zer = K.zeros((2,3,4))
```

The `K.zeros_like` still outputs a tensor, but `K.zeros` creates a Variable:

```
In [11]: k_zer_like
Out[11]: <tf.Tensor 'zeros_like_1:0' shape=(2, 3, 4) dtype=float32>

In [12]: k_zer
Out[12]: <tensorflow.python.ops.variables.Variable at 0x7f5969ab2890>

In [13]: print tf.get_default_graph().get_collection('variables')
[<tensorflow.python.ops.variables.Variable object at 0x7f5969ab2890>]
```

To avoid Variable creation, we can create `initial_states` using only the `K.zeros_like` call. Alternately Keras could expose the `tf.zeros` method somehow, but would be a larger design decision.
